# Before class

* Make sure you have downloaded the 'musical_instrument_reviews.csv' data file. This is a dataset containing reviews for a product on Amazon (taken from [a larger dataset on Kaggle](https://www.kaggle.com/eswarchandt/amazon-music-reviews)), and we'll be cleaning the data so we can see if customers liked the product or not.


# Outline of class agenda

During this class you'll:

1. Learn how to select data in DataFrames by columns and by rows
2. Learn how to change DataFrame column names
3. Learn how to make new columns and remove columns
4. Learn how to spot missing data in DataFrames and replace it
5. Learn how to sort data in DataFrames
6. Learn how to export data from a DataFrame back to a csv file
7. Get comfortable with the 'data cleaning' process in python overall
 

# Class

Today we're going to focus on 'data cleaning'. But, what is data cleaning, exactly?
* Any operations you need to do to get your data in shape for analysis, graphs, or presentations
* Typically involves checking the data for correctness, fixing errors, and dealing with missing data
* Might involve selecting only certain parts of larger datasets, or combining multiple datasets

Data cleaning might not always be the most exciting aspect of working with data, but it is *super important* for any projects where you're working with data! It also is something most data scientists spend a [lot of time](https://businessoverbroadway.com/2019/02/19/how-do-data-professionals-spend-their-time-on-data-science-projects/) on


To get started for today, we'll import pandas as usual, then read in a daset on musical instruments from a .csv file

First import pandas

In [186]:
import pandas as pd

Then, read in the csv file using `pd.read_csv()`

In [187]:
reviews = pd.read_csv('../../datasets/musical_instrument_reviews.csv')


Now, let's take a quick look at it:

In [188]:
reviews.head()

reviewerID                                      reviewerName  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...   
1  A14VAT5EAX3D9S                                              Jake   
2  A195EZSQDW3E21                     Rick Bennette "Rick Bennette"   
3  A2C00NNG1ZQQG2                         RustyBill "Sunday Rocker"   
4   A94QU4C90B1AX                                     SEAN MASLANKA   

                                          reviewText  overall  \
0  Not much to write about here, but it does exac...      5.0   
1  The product does exactly as it should and is q...      5.0   
2  The primary job of this device is to block the...      5.0   
3  Nice windscreen protects my MXL mic and preven...      5.0   
4  This pop filter is great. It looks and perform...      5.0   

                                 summary   reviewTime  reviewWords  
0                                   good  02 28, 2014           51  
1                                   Jake  03 16, 2013          104  
2                   It Does The Job Well  08 28, 2013           77  
3          GOOD WINDSCREEN FOR THE MONEY  02 14, 2014           35  
4  No more pops when I record my vocals.  02 21, 2014           28

Before we get started on data cleaning, we can check out a few things about this DataFrame. This contains 500 reviews of a product on Amazon (a screen for a microphone to help with music recording), so with this analysis we can imagine we're trying to figure out whether people seem to like this product, or whether there are certain reasons why people give it negative reviews. 

We have several pieces of data on each review, stored in each column:
* **reviewerID:** a unique ID for each Amazon user who reviewed this product
* **reviewerName:** the username for each reviewer
* **reviewText:** a string containing the text of the entire review
* **overall:** the user's overall rating of the product on a scale from 1-5 stars
* **summary:** the user's summary of their review
* **reviewTime:** the date of the review [day month year]
* **reviewWords:** the number of words in the user's entire review (i.e. number of words in the entry in the `reviewText` column)

## 1. Selecting data in a DataFrame by columns and by rows

Last class, we learned how to select specific cells of DataFrame objects using numeric indexing with `iloc()`, but often it is much more useful to be able to work with columns and rows in different ways. We'll learn a few of those here:

### Selecting columns

There are two methods that are useful for selecting certain columns in a DataFrame.

**Method 1**

The first is using brackets with the column labels inside. This method is really flexible, because we can either select just one column as a Pandas Series object

`reviews['reviewerID']`

Or, we can one or more columns as an entire DataFrame object by using two sets of brackets (so the column labels are a list object


`reviews[['reviewerID']]`

`reviews[['reviewerID', 'overall']]`

**Method 2**

The second method is using `.` to reference a column as an attribute of a DataFrame. This always only gets 1 column of the data frame as a Pandas Series:

`reviews.reviewerID` (this is equivalent to `reviews['reviewerID']`)

The only thing to be aware of for Method 2 is that you can't make *new* columns this way. Colummns referenced with the `.` have to already exist in the DataFrame

**Mini-challenge:** 
1. Select the `overall` column of the `reviews` DataFrame, and save if to the variable 'overall_rating'. Then run `type()` to see what kind of object you get.
2. Make a new DataFrame of the columns `overall` and `reviewWords` and save it into a variable called `df2`. Confirm that it is a DataFrame using `type()`

**Solution:** 


In [189]:
# part 1
overall_rating = reviews.overall
type(overall_rating)

pandas.core.series.Series

In [190]:
# part 2
df2 = reviews[['overall', 'reviewWords']]
type(df2)

pandas.core.frame.DataFrame

One more quick trick is that you can also make a list of column names, and then use this list to create a DataFrame with only these columns, for example:

In [191]:
column_list = ['overall', 'reviewText', 'summary']
df3 = reviews[column_list]
df3.head()

overall                                         reviewText  \
0      5.0  Not much to write about here, but it does exac...   
1      5.0  The product does exactly as it should and is q...   
2      5.0  The primary job of this device is to block the...   
3      5.0  Nice windscreen protects my MXL mic and preven...   
4      5.0  This pop filter is great. It looks and perform...   

                                 summary  
0                                   good  
1                                   Jake  
2                   It Does The Job Well  
3          GOOD WINDSCREEN FOR THE MONEY  
4  No more pops when I record my vocals.

Because `column_list` in this example is already a list, we only need 1 set of brackets here to make a DataFrame with these specified columns

### Selecting or 'filtering' rows (based on their content)

So, we've learned how to select only certain columns from a DataFrame. Great! 

One other thing that might be really helpful is only selecting certain *rows*. We saw how to do that last time with `iloc()`, but what might often be more useful is only including rows that with data in them that *fit certain criteria*. Often, we might call this **filtering** the data.

For example, when we're going through these product reviews, we might want to 
* filter the data for reviews that were longer than 50 words
* filter the data for reviews in the year 2014

Generally, we filter rows by applying a [conditional statement](https://www.python-course.eu/python3_conditional_statements.php#:~:text=statements%20in%20Python.-,Conditional%20statements%20in%20Python,should%20be%20executed%20or%20not.&text=The%20indented%20block%20is%20only,condition%20'condition'%20is%20True.) (a logical statement that can return `True` or `False` for each entry) related to a column of the DataFrame

For example, one conditional statement to check if reviews are longer than 50 words could be:

In [192]:
reviews.reviewWords > 50

0       True
1       True
2       True
3      False
4      False
       ...  
495    False
496    False
497     True
498    False
499    False
Name: reviewWords, Length: 500, dtype: bool

If we do the conditional statement *by itself* like this, it returns a pandas Series full of `True` and `False` values. For each review, we get `True` if the condition is met (the review was longer than 50 words, or `reviewWords > 50`) and `False` if not.

Now, the *fancy part* comes in when we use this Series of `True` and `False` values to select which rows of whole DataFrame we want. We can use this to take **only the rows of the DataFrame where the condition is met, where the value is `True`** with the following syntax:

In [193]:
reviews_above_50_words = reviews[reviews.reviewWords > 50]

This is a *bunch* of steps in one line of code, but to summarise, what this does is to apply a **condition** to the `reviewWords` column of the DataFrame, which only returns `True` if the numeric value for an entry in that column is above 50. Then, the bracket indexing selects *only* the rows with a value of `True` from `reviews`, and we save that into a new DataFrame object. Let's take a look to confirm that we're only getting rows where the review is over 50 words:

In [194]:
reviews_above_50_words.head()

reviewerID                                      reviewerName  \
0   A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...   
1   A14VAT5EAX3D9S                                              Jake   
2   A195EZSQDW3E21                     Rick Bennette "Rick Bennette"   
7    AJNFQI3YR6XJ5                                 Fender Guy "Rick"   
10  A2NYK9KWFMJV4Y                       Mike Tarrani "Jazz Drummer"   

                                           reviewText  overall  \
0   Not much to write about here, but it does exac...      5.0   
1   The product does exactly as it should and is q...      5.0   
2   The primary job of this device is to block the...      5.0   
7   I now use this cable to run from the output of...      3.0   
10  Monster makes a wide array of cables, includin...      5.0   

                                              summary   reviewTime  \
0                                                good  02 28, 2014   
1                                                Jake  03 16, 2013   
2                                It Does The Job Well  08 28, 2013   
7                  Didn't fit my 1996 Fender Strat...  11 16, 2012   
10  One of the best instrument cables within the b...  04 19, 2012   

    reviewWords  
0            51  
1           104  
2            77  
7           167  
10          190

**Mini-challenge:**

In a similar way to what we just did, filter `reviews` to make a new DataFrame called `good_reviews` with only `overall` scores of 4 or higher.

**Solution**

In [195]:
good_reviews = reviews[reviews.overall >= 4]

We can use the describe() function to check that the minumum value in the overall column is a 4

In [196]:
good_reviews.describe()

overall  reviewWords
count  451.000000   451.000000
mean     4.725055    86.119734
std      0.446982   107.781997
min      4.000000     7.000000
25%      4.000000    30.000000
50%      5.000000    49.000000
75%      5.000000    90.500000
max      5.000000   915.000000

## 2. Changing DataFrame column names

To update the column names, we can use the [`rename()` function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html). To specify which columns we want to rename and what the new names should be, we add a `columns` argument in the form of a dictionary. In this dictionary object, each column name is formatted as `{'old name':'new name'}`

So, if we wanted to change the `summary` column to be called `review_summary`:

In [197]:
reviews.rename(columns = {'summary':'review_summary'}, inplace = True)

Here, we also can add the `inplace = True` arguement so that the column is renamed 'inplace' and the DataFrame is modified without us having to assign it to a variable. The colum name is now changed:

In [198]:
reviews.head(3)

reviewerID                                      reviewerName  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...   
1  A14VAT5EAX3D9S                                              Jake   
2  A195EZSQDW3E21                     Rick Bennette "Rick Bennette"   

                                          reviewText  overall  \
0  Not much to write about here, but it does exac...      5.0   
1  The product does exactly as it should and is q...      5.0   
2  The primary job of this device is to block the...      5.0   

         review_summary   reviewTime  reviewWords  
0                  good  02 28, 2014           51  
1                  Jake  03 16, 2013          104  
2  It Does The Job Well  08 28, 2013           77

We can also change multiple columns at once with the same syntax:

In [199]:
reviews.rename(columns = {'reviewerName':'username',
                          'reviewerID': 'user_id'}, inplace = True)

In [200]:
reviews.head(3)

user_id                                          username  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...   
1  A14VAT5EAX3D9S                                              Jake   
2  A195EZSQDW3E21                     Rick Bennette "Rick Bennette"   

                                          reviewText  overall  \
0  Not much to write about here, but it does exac...      5.0   
1  The product does exactly as it should and is q...      5.0   
2  The primary job of this device is to block the...      5.0   

         review_summary   reviewTime  reviewWords  
0                  good  02 28, 2014           51  
1                  Jake  03 16, 2013          104  
2  It Does The Job Well  08 28, 2013           77

## 3. Adding and removing DataFrame columns

Often we'll want to add new columns to our DataFrame to make new useful variables, or remove columns we're not using. Let's learn how to do both here!

**Adding a new column**

In general, we can add new columns using the `df['column_name']` syntax just as if we were working with existing columns. 

As an example, maybe we want to make a new column that indicates if a review was especially long (over 500 words) called `long_review`. 

In [201]:
reviews['long_review'] = reviews.reviewWords > 100

We can also make columns where all of the values are exactly the same by assigning an integer, float, string, or boolean value to a column:

In [202]:
# make a column indicating the name of the data analyst
reviews['data_analyst'] = 'Paul Bloom'

# make a column with all zeros
reviews['zeros'] = 0

Now we can check that those columns are there:

In [203]:
reviews.head(3)

user_id                                          username  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...   
1  A14VAT5EAX3D9S                                              Jake   
2  A195EZSQDW3E21                     Rick Bennette "Rick Bennette"   

                                          reviewText  overall  \
0  Not much to write about here, but it does exac...      5.0   
1  The product does exactly as it should and is q...      5.0   
2  The primary job of this device is to block the...      5.0   

         review_summary   reviewTime  reviewWords  long_review data_analyst  \
0                  good  02 28, 2014           51        False   Paul Bloom   
1                  Jake  03 16, 2013          104         True   Paul Bloom   
2  It Does The Job Well  08 28, 2013           77        False   Paul Bloom   

   zeros  
0      0  
1      0  
2      0

**Removing Columns**

We can remove columns using the `drop()` function as follows:

In [204]:
reviews.drop(columns = ['data_analyst', 'zeros'], inplace = True)

So here with `drop()`, we specify a `columns` argument as a list of the column names we want to remove. If we only wanted to drop one column, we could also use a list with just 1 element.

You might wonder be wondering what `inplace = True` does. This argument saves the DataFrame with the columns dropped *back* into the original one, so we don't have to reassign it to a variable again. If we don't include this, we'd have to assign to a variable to save the changes.

If you have a DataFrame with a LOT of columns and you want to get rid of most of them, it is often easier to select the columns you *do* want rather than the ones you *don't*. (see #1 for selecting columns)

**Mini-challenge**

* Add a column called 'my_name' to `reviews` with your name in each row
* Add a column called 'my_integer' to `reviews` with the same integer in each row
* Then, remove both columns uding `inplace = True`

**Solution**

In [205]:
# adding the columns
reviews['my_name'] = 'Paul Bloom'
reviews['my_integer'] = 8
reviews.head(2)

user_id                                          username  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...   
1  A14VAT5EAX3D9S                                              Jake   

                                          reviewText  overall review_summary  \
0  Not much to write about here, but it does exac...      5.0           good   
1  The product does exactly as it should and is q...      5.0           Jake   

    reviewTime  reviewWords  long_review     my_name  my_integer  
0  02 28, 2014           51        False  Paul Bloom           8  
1  03 16, 2013          104         True  Paul Bloom           8

In [206]:
# drop the new columns
reviews.drop(columns=['my_name', 'my_integer'], inplace = True)

In [207]:
reviews.head(2)

user_id                                          username  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...   
1  A14VAT5EAX3D9S                                              Jake   

                                          reviewText  overall review_summary  \
0  Not much to write about here, but it does exac...      5.0           good   
1  The product does exactly as it should and is q...      5.0           Jake   

    reviewTime  reviewWords  long_review  
0  02 28, 2014           51        False  
1  03 16, 2013          104         True

## 4. Finding & dealing with missing data

Often the data we're working with has missing values we'll need to deal with before analysis or graphing. 

One way to check for these values in the entire DataFrame is with the `isnull()` function. For example, if we call this on the entire dataframe, the function returns a DataFrame where each cell is `True` if that same cell in the original dataframe contained a null (or missing) value, and `False` if there was data in that cell.

In [208]:
reviews.isnull()

user_id  username  reviewText  overall  review_summary  reviewTime  \
0      False     False       False    False           False       False   
1      False     False       False    False           False       False   
2      False     False       False    False           False       False   
3      False     False       False    False           False       False   
4      False     False       False    False           False       False   
..       ...       ...         ...      ...             ...         ...   
495    False     False       False    False           False       False   
496    False     False       False    False           False       False   
497    False     False       False    False           False       False   
498    False     False       False    False           False       False   
499    False     False       False    False           False       False   

     reviewWords  long_review  
0          False        False  
1          False        False  
2          False        False  
3          False        False  
4          False        False  
..           ...          ...  
495        False        False  
496        False        False  
497        False        False  
498        False        False  
499        False        False  

[500 rows x 8 columns]

This might be hard to read by itself, but then if we combine this with the `sum()` function, we can get a quick summary of how many missing values are in each column:

In [209]:
reviews.isnull().sum()

user_id           0
username          4
reviewText        0
overall           4
review_summary    3
reviewTime        0
reviewWords       0
long_review       0
dtype: int64

Here we can see that the `username` and `overall` columns are both missing 4 values, and `review_summary` is missing 3

But, what if we wanted to actually check out what's going on in the rows missing data? For this, we can use the same process we used earlier to *select rows*, but here, we can select the rows where a certain column contains missing values. For example, let's select only the rows where `overall` is null:

In [210]:
# This is the same process as in section 1! Inside the brackets is a conditional argument
# The code inside the brackets gives True for every null value in the column `overall` and false otherwise
# Then we filter the whole DataFrame for only the rows where this returned True
reviews[reviews.overall.isnull()]

user_id    username  \
137   AFLRU6952DEFX          S.   
140   AQFOCVEBDCYU9  Jazzgryl52   
245  A37AQI4AU3JWSR      Joshua   
246   AUK79PXTAOJP9      ~ Kyle   

                                            reviewText  overall  \
137  These have a lower profile and can fit and ben...      NaN   
140  I purchased this to work with my ION system an...      NaN   
245  Donr be fooled by the imitations... should be ...      NaN   
246  Great rack mount screws. Rubber washers are pe...      NaN   

                                review_summary   reviewTime  reviewWords  \
137  Possibly better than the metal ended ones  06 24, 2014           20   
140                                No problems  07 31, 2013           22   
245           Best rack screws for your money.  12 18, 2012           28   
246                                      Great   07 8, 2013           44   

     long_review  
137        False  
140        False  
245        False  
246        False

**Removing missing values**

A lot of the time we might want to remove rows where a certain column is missing data. This will often show up as `NaN`, which stands for 'not a number'. 

Here, we can't really do much without the `overall` ratings of the product, so we'll remove the 4 rows with NaNs in this column. We can do this by putting a minus sign in front of the `reviews.overal.isnull()` which will return True for all **non-null** rows. Then, we select only those, and save into a new DataFrame called `reviews_no_null_ratings`

In [211]:
reviews_no_null_ratings = reviews[-reviews.overall.isnull()]

Now, this new DataFrame has no nulls in the `overall` column, and we can see it is 4 rows smaller than before (only 496 columns now)

In [212]:
print(reviews_no_null_ratings.isnull().sum())
print(reviews_no_null_ratings.shape)

user_id           0
username          4
reviewText        0
overall           0
review_summary    3
reviewTime        0
reviewWords       0
long_review       0
dtype: int64
(496, 8)


**Replacing Missing Values**

Instead of removing data, sometimes we might want to replace it with something meaningful. 

Here, we want to replace the NaN values in the `username` column with the username 'unknown user'. First, we can take a look at these:

In [213]:
reviews[reviews.username.isnull()]

user_id username  \
41    AA5TINW2RJ195      NaN   
263  A14VAT5EAX3D9S      NaN   
264   A2RVY2GDMZHH4      NaN   
432  A3BMYEA3J6RBVV      NaN   

                                            reviewText  overall  \
41             Good quality cable and sounds very good      5.0   
263  It is exactly what you need in a capo! You can...      5.0   
264  Love these capos and nice that you can get the...      5.0   
432    Bought it as a gift. Friend loved it very much.      5.0   

                review_summary   reviewTime  reviewWords  long_review  
41                  Five Stars  07 15, 2014            7        False  
263                      Great   04 2, 2013           82        False  
264         Always great Kyser  12 28, 2012           23        False  
432  Friend loved it very much  07 13, 2014           10        False

Now, we use these to reassign values to `reviews.username` 

In [227]:
reviews.username[reviews.username.isnull()] = 'unknown user'

/Users/paul/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Now if we look for nulls in this column, there aren't any:

In [224]:
# returns no data!
reviews[reviews.username.isnull()]

Empty DataFrame
Columns: [user_id, username, reviewText, overall, review_summary, reviewTime, reviewWords, long_review]
Index: []

But, if we look for values of 'unknown user' in this column, we find them where the NaNs used to be 

In [225]:
reviews[reviews.username == 'unknown user']

user_id      username  \
41    AA5TINW2RJ195  unknown user   
263  A14VAT5EAX3D9S  unknown user   
264   A2RVY2GDMZHH4  unknown user   
432  A3BMYEA3J6RBVV  unknown user   

                                            reviewText  overall  \
41             Good quality cable and sounds very good      5.0   
263  It is exactly what you need in a capo! You can...      5.0   
264  Love these capos and nice that you can get the...      5.0   
432    Bought it as a gift. Friend loved it very much.      5.0   

                review_summary   reviewTime  reviewWords  long_review  
41                  Five Stars  07 15, 2014            7        False  
263                      Great   04 2, 2013           82        False  
264         Always great Kyser  12 28, 2012           23        False  
432  Friend loved it very much  07 13, 2014           10        False

**Removing ALL missing values**

So far we've been targeting missing values in each column one-by-one. But, what if we wanted to remove ALL the missing data from the DataFrame in one step? We can use the `dropna()` function for this!

If we want to remove each *row* that has any missing values, we can run as follows:

In [231]:
reviews_clean = reviews.dropna(axis = 0)

Here `axis = 0` specifies that we removing rows (because rows are axis 0 and columns are axis 1)

In [232]:
reviews_clean.isnull().sum()

user_id           0
username          0
reviewText        0
overall           0
review_summary    0
reviewTime        0
reviewWords       0
long_review       0
dtype: int64

## Topic 2


Here is a great sentence with some text for the lesson, as well as inline code like `print('hello world)` or `git status`

# Challenge

For this challenge in this class, make an awesome python function!